OpenRA.Game/Map/CellLayer.cs

```diff
@@ -45,9 +45,23 @@ public void CopyValuesFrom(CellLayer<T> anotherLayer)
 			Array.Copy(anotherLayer.entries, entries, entries.Length);
 		}
 
+		public static CellLayer<T> CreatePrototype(Func<int, int, T> creator, int mapWidth, int mapHeight, TileShape tileShape)
+		{
+			var size = new Size(mapWidth, mapHeight);
+			var defaultCellInfoLayer = new CellLayer<T>(tileShape, size);
+			for (var v = 0; v < mapHeight; v++)
+				for (var u = 0; u < mapWidth; u++)
+					defaultCellInfoLayer[new MPos(u, v)] = creator(u, v);
+
+			return defaultCellInfoLayer;
+		}
+
 		// Resolve an array index from cell coordinates
 		int Index(CPos cell)
 		{
+			if (Shape == TileShape.Rectangle)
+				return cell.Y * Size.Width + cell.X;
```

Comment:
Any time a cellLayer position was accessed by CPos, the performance improvement I could experiment with that simple "hack" was almost x2 according to some profiling information.

But as I tell you, this is a "soft hack". The SRP (Single Responsibility Principle) is systematically violated in OpenRA code. You can see how I could break the Pathfinder into roughly 10 smaller, more organized classes.

Maybe one cool workaround could be implement an implicit operator caster inside CPos or MPos. This is something to study.